### DISCUSSION TITLE
Welcome to osm4scala Discussions!

### DISCUSSION THREAD

**Discussion by angelcervera** [simplexspatial/osm4scala#77] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with *osm4scala* users. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we build together 💪.

> :warning: For Q&A, please, use [StackOverflow](https://stackoverflow.com/tags/osm4scala) tag [osm4scala].

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by shantanuaggarwal2695** [comment]
Hello

Can we fetch polygonal coordinates for relation types? I have a use case where I need to run various spatial queries on various types of geometries and for that, I require polygonal coordinates for ways and relations.

**Reply by angelcervera** [reply]
Hi @shantanuaggarwal2695 
As part of the pbf specs, _Relations_, like _Ways_, do not have coordinates inside. They contain IDs of _Nodes_.

To be able to do it, you will need to cross nodes and relations. The best way is to use the Spark connector to write a query.

Maybe what you a looking for is a way to translate the PBF into a set of GeoJSON objects or another format that contains coordinates in a Line, Polygon, etc.. object. Am I right?

BTW. Maybe this conversation should be in a different thread.

**Reply by shantanuaggarwal2695** [reply]
Hi @angelcervera 

Yes, I am looking to construct vector objects from given nodes, relations, and ways and for that, I need to have lat-long coordinates of polygons and points. So should I cross nodes and relations in order to fetch geometries?

**Reply by angelcervera** [reply]
Hi @shantanuaggarwal2695 
The target of the Spark connector is to enable this kind of operation on top of PBF files.
So yes, my recommendation is to cross Node entities and Relation entities to generate the format that you wish.
You _should_ be able to execute it even using a laptop with 16GB ram for the full planet in one/two hour process.
I did it few times, usually for Ways. Basically:
1. Extract Nodes.
2. Extract Relations, keeping the node order.
3. Join aggregating by relation id.
4. Map the result to the output format that you need.

To do it over Relations, it is a little bit more difficult because every element in the Relation could be a Way, a Node or another Relation.

**Comment by Kazerujoy** [comment]
Hello 
I try to read osm files by osm4scala in spark with maven, i add the dependence in maven and do the spark.read.format('osm.pbf').load(). The exception say there Failed to find data source:osm.pbf. how can i deal with this problem?

**Reply by angelcervera** [reply]
This usually happens if you did not add the jar properly.
Could you share more information about how are you doing it? Are you using it in a notebook or in an application? Are you using the last version?
Are you creating an uber jar to create your jar or you are using the `--packages` parameter to add the dependency when you submit the spark job?

In the [Documentation](https://simplexspatial.github.io/osm4scala/docs/spark-connector) you can find step-by-step information about how to do it in different contexts.

Also, you can find [examples in source code](https://github.com/simplexspatial/osm4scala/tree/master/examples/spark-utilities).

BTW: Could we move this conversation to Stackoverflow?

**Reply by angelcervera** [reply]
I found that you closed ticket #90 Did you find the issue?

**Reply by Kazerujoy** [reply]
I add osm4scala by add maven dependence in my project's pom.xml.
And i download the jar and add it to my lib, now it works, thx a lot.
Not good in using github, lol
